### PR TITLE
Digital GPIOs values clamped 0x0 or 0x1. Added Input Data Register (IDR) and Output Data Register feature (ODR).

### PR DESCRIPTION
**Pull request by stfwi**
Hi, @frodare, I played along a bit with additional use cases of the mod and addressed the digital mode value range in one go.
The main change is adding the two registers IDR and ODR to read/write the bit encoded data words of all I/Os. That could be interesting to latch inputs as a consistent state of one tick before processing.

Also extended the manual as I have the impression people do read this prior to checking the wiki pages and need some kind of brief introduction in-game. Now, I am not a native EN, so I would feel relieved if you could risk a glimpse for typos or strange formulations.

Keep it up, cheers, wile.

### CODE REVIEW

File: src/main/java/net/torocraft/minecoprocessors/util/BookCreator.java [hunk 1]
```diff
@@ -1,11 +1,10 @@
 package net.torocraft.minecoprocessors.util;
 
-import java.io.BufferedReader;
-import java.io.FileNotFoundException;
-import java.io.IOException;
-import java.io.InputStream;
-import java.io.InputStreamReader;
```
Comment (paperky) [inline src/main/java/net/torocraft/minecoprocessors/util/BookCreator.java:7]: Does replacing the individual imports with `io.*` result in more components being imported?

I'd assume that `java.io` contains more than just `BufferedReader`, `FileNotFoundException`, `IOException`, `InputStream`, and `InputStreamReader`, and that `import java.io.*` imports all classes as well. Would there be downside to this?

File: src/main/resources/data/minecoprocessors/books/manual.txt [hunk 2]
```diff
@@ -1,130 +1,291 @@
...
+
+assigns 10 to the register §1a§r ("§1a = 10§r").
+~~~
+Instructions which calculate something normally store the result back in the first argument, hence,
+
+ §1MOV a, 12§r
+ §1MOV b, 10§r
+ §1SUB a, b§r
+
+will result in register §1a§r containing §12§r as present value.
```
Comment (paperky) [inline src/main/resources/data/minecoprocessors/books/manual.txt:139]: I'd reccommend changing this to read
`will result in register §1a§r containing the value §12§r.`

File: src/main/resources/data/minecoprocessors/books/manual.txt [hunk 3]
```diff
@@ -1,130 +1,291 @@
...
-§lWFE§r Sleep (experimental)
+The CPU interacts with the world around via special registers. These are mapped to the four sides of the device, and we need to tell the CPU if we like to read signals from a side, or want to output redstone power there.
+
+Additionally, we can
 ~~~
-§1Arithmetic Instructions§r
+specify for each I/O if we are interested in the numeric "strength" value of the signal, or only the on/off data.
+§oAlso known as §1analog§r and §1digital§r mode.§r
+
+Lastly, ports can also be used to reset the controller by leaving a side unconfigured or "unconfiguring" it explicitly.
```
Comment (paperky) [inline src/main/resources/data/minecoprocessors/books/manual.txt:179]: This seems to at least gloss over or maybe even misinterpret how reset ports are shown to work on the Wiki:
Ports are used as reset ports only if the corresponding bit in the high nybble of PORTS is set, and the port is currently set to input mode, according to [this](https://github.com/ToroCraft/Minecoprocessors/wiki/I-O-Ports) page on the wiki

File: src/main/resources/data/minecoprocessors/books/manual.txt [hunk 4]
```diff
@@ -1,130 +1,291 @@
...
+
+§4start:§r
+ §1JMP§r §4start§r
+~~~
+Numbers can be hex, decimal, octal, as well as binary. A byte with all bits set (§11§r) can be written as:
+
+Hexadecimal: §10xff§r
+Decimal: §1-1§r
+Octal: §10o377§r
+Binary: §111111111b§r
```
Comment (paperky) [inline src/main/resources/data/minecoprocessors/books/manual.txt:156]: Is it ever possible to actually use the full 8bit binary representation of a number in a program? From using the mod, I remember there being a hard limit on the horizontal characters on a line, and I'm not aware of any line continuation characters, so I don't think it's possible. Maybe that should be addressed somewhere, that you can't write out all 8 binary digits in-program?

File: src/main/resources/data/minecoprocessors/books/manual.txt (hunk 3 above)
Comment (paperky) [reply src/main/resources/data/minecoprocessors/books/manual.txt:179]: Looks like this is mentioned in lines +196 and +198. Maybe change the wording here to say something like
`Finally, any of the ports can also be used as a reset for the processor by setting the correct bits in the ports register`

File: src/main/resources/data/minecoprocessors/books/manual.txt [hunk 5]
```diff
@@ -1,130 +1,291 @@
...
 
-§lSHL a§r Shift Left
-§lSHR a§r Shift Right
-§lSAR a§r Arithmetic Right
-§lSAL a§r Arithmetic Left
-§lROR a§r Rotate Right
-§lROL a§r Rotate Right
+  §1MOV ports, 1010b§r
+
+defines §1pf§r and §1pl§r as outputs, §1pb§r and §1pr§r as inputs. To use an I/O for reset, set its bit
```
Comment (paperky) [inline src/main/resources/data/minecoprocessors/books/manual.txt:195]: Consider changing `To use an I/O for reset,` to `To use an I/O port for reset,` for clarity and consistiency

File: src/main/resources/data/minecoprocessors/books/manual.txt [hunk 6]
```diff
@@ -1,130 +1,291 @@
...
-§lJNC label§r if no Carry
-§lDJNZ a, label§r Dec and Jump if not zero
-§lCALL label§r Call
-§lRET§r Return
-§lCMP a, b§r Compare
+By default I/Os are in digital mode. Hence, we read the value §10x00§r if an input signal is §40§r, and §10x01§r for any
 ~~~
-§1Number Formats§r
+strength from §41§r to §415§r.
+Reversely, writing §10§r to an output will result in an unpowered side, while any other value will yield strength §415§r.
```
Comment (paperky) [inline src/main/resources/data/minecoprocessors/books/manual.txt:206]: Consider changing this wording:
`Similarly, writing §10§r to an output port will yield strength §40§r, on that side, and any other value will yield signal strength §415§r.`

I admit that there might not be space on the page left for this longer line, but I'm not able to check that currently

File: src/main/resources/data/minecoprocessors/books/manual.txt [hunk 7]
```diff
@@ -1,130 +1,291 @@
...
-§l-1§r two's complement
-§l0xff§r Hexadecimal
-§l0o377§r Octal
-§l11111111b§r Binary
+  §1MOV pf, 0§r
+  §1MOV pf, 1b§r
+  §1MOV pf, 0x0e§r
+  §1MOV pf, -3§r
+
+Signal §40§r in the first case, others §415§r.
```
Comment (paperky) [inline src/main/resources/data/minecoprocessors/books/manual.txt:213]: Consider changing the wording to:
`If the front port is set as an output, then these instructions would set it to`
`signal strength §40§r, for the first instruction, and §415§r for the others.`

File: src/main/resources/data/minecoprocessors/books/manual.txt [hunk 8]
```diff
@@ -1,130 +1,291 @@
...
 ~~~
-§1Registers§r
+Multiple digital I/Os can be read/written simultaneously using Input/Output Data Registers §1idr§r and §1odr§r:
+
+  §1MOV a, idr§r
+  §1MOV odr, 0110b§r
+
+The §1idr§r reads set bits of output ports as 1. These registers can be useful to latch ports in one tick, and
```
Comment (paperky) [inline src/main/resources/data/minecoprocessors/books/manual.txt:220]: Might be good to mention the behavior of `idr` and `odr` when any of the ADC bits are set

File: src/main/java/net/torocraft/minecoprocessors/util/BookCreator.java (hunk 1 above)
Comment (stfwi) [reply src/main/java/net/torocraft/minecoprocessors/util/BookCreator.java:7]: Just quickly double checked, IJ automatically replaced the single imports again with `java.io.*`, I presume it's not crucial in Java (I am no expert at all in J). Gonna push an updated version, explicit includes are generally better.

File: src/main/resources/data/minecoprocessors/books/manual.txt (hunk 4 above)
Comment (stfwi) [reply src/main/resources/data/minecoprocessors/books/manual.txt:156]: Increased to 22 chars per line with the last push. Maybe the bode book ui width needs to be adapted to that eventually, if I will the line with "A"s it's over the writable range - there are worse things in the world however ;)

**Comment by stfwi** [comment]
With the manual I need to check at the weekend - did you check if the words still fit on the pages or wrap?
(The change in the book creator was to edit the manual while the game is running, it looks for the "manual.txt" in the game instance root folder, so closing and re-opening the book will immediately reload the changes). 
If you have time and like to check that, I would be glad. Cheers.

File: src/main/resources/data/minecoprocessors/books/manual.txt (hunk 6 above)
Comment (paperky) [reply src/main/resources/data/minecoprocessors/books/manual.txt:206]: Typo of a comma, and maybe change to
`Similarly, writing §10§r to a digital output port will yield strength §40§r, and any other value will yield signal strength §415§r.`

File: src/main/java/net/torocraft/minecoprocessors/processor/Register.java [hunk 9]
```diff
@@ -52,5 +52,15 @@
    *
    * <ul> <li><b>bit 0:</b> front port</li> <li><b>bit 1:</b> back port</li> <li><b>bit 2:</b> left port</li> <li><b>bit 3:</b> right port</li> </ul>
    */
-  ADC
+  ADC,
+  /**
+   * Digital Input Data Register (reads all input ports as nibble).
+   * Bit order: (pf<<0)|(pb<<1)|(pl<<2)|(pr<<3)
+   */
+  IDR,
```
Comment (frodare) [inline src/main/java/net/torocraft/minecoprocessors/processor/Register.java:60]: Interesting idea.  How do you handle conflicts between IDR or ODR and PF/PR/PL/PB?

Comment (stfwi) [reply src/main/java/net/torocraft/minecoprocessors/processor/Register.java:60]: Hi, based on the assumption of the one-instruction-per-tick I synchronise IDR while reading the inputs (`updateInputPorts()`), and ODR while writing the outputs (`updateOutputs()`). pf,pb, etc have an edge detection to firing `neighbourChanged` on adjacent blocks only when appropriate, and this was done for the ODR, too. So for one tick, only one of `odr` or ports should change if I did not manage to build in a late night bug ;). Cheers,-

**Comment by frodare** [comment]
@stfwi I am not planning on working on this mod past the 1.12 master branch.  Since you have done all the work getting the new versions working, I will let you decide bug fixes and features make sense.  Also, please feel free to fork and rebrand the project and take it in what ever direction you want. I don't think I will be working on this in future versions and I don't want to slow your progress down.  

Let me know if this PR is ready to be built.  If so I will try to get it out this week.

**Comment by stfwi** [comment]
Hi @frodare, sorry about the late response. There are some open manual proposals by @paperky that I would like to address, but I am currently a bit short on time. Would it be acceptable to push this one up as beta and then do one release version when this is manual sorted out? I can build (and sign) the jar with the corresponding changelog if you like and attach it.

Thanks for the offer to continue the project independently for >1.12 versions. I am honestly not sure how to deal with that, as the mod is somehow mentally linked to ToroCraft in my head. As there are no massive changes planned (the IDR/ODR PR is only to enable latching all ports in one tick for a consistent dataset), I could also imagine to provide the few maintenance updates/builds for the ToroCraft Team without being mentioned, registered, or so. It's a hobby for me, and this mod is simply one of the good mods with educational value, to my brain totally underrated, and an invention in MC worth preserving (not only for "Look lad, this is how we had to code in year 19...", but more for "Look lad, this is how a machine sees it"). If I could I would to a spotlight or so.

Long tldr paragraph - Let me know what you think. Best wishes -

**Pull request merged**